src/dinitctl.cc

```diff
@@ -1287,3 +1304,47 @@ static int enable_disable_service(int socknum, cpbuffer_t &rbuffer, const char *
 
     return 0;
 }
+
+static int do_setenv(int socknum, cpbuffer_t &rbuffer, char **env_names, bool verbose)
+{
+    using namespace std;
+
+    string buf;
+
+    while (*env_names) {
+        const char *envp = *env_names++;
+        buf.clear();
+        buf.reserve(6);
+        // protocol message and size space
+        buf.push_back(DINIT_CP_SETENV);
+        buf.append(2, 0);
+        // either full var or name
+        auto elen = strlen(envp);
+        buf.append(envp, elen);
+        // = not found, get value from environment
+        if (!memchr(envp, '=', elen)) {
+            buf.push_back('=');
+            auto *envv = getenv(envp);
+            if (envv) {
+                buf.append(envv);
+            }
+        }
+        uint16_t bufs = buf.size() - 3;
+        // set size in protocol message
+        memcpy(&buf[1], &bufs, 2);
+        // send
+        write_all_x(socknum, buf.data(), bufs + 3);
+        wait_for_reply(rbuffer, socknum);
+        if (rbuffer[0] == DINIT_RP_BADREQ) {
+            if (verbose) {
```

Comment:
Don't need `verbose` check for error messages, error output should always visible